org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/CompilationUnitScope.java

```diff
@@ -240,10 +241,34 @@ void checkAndSetImports() {
 	resolvedImports[0] = getDefaultImports()[0];
 	int index = 1;
 
+	Predicate<ImportReference> isStaticImport = i -> i.isStatic();
+	Predicate<ImportReference> isNotStaticImport = Predicate.not(isStaticImport);
+
+	// GitHub 269: resolve non-static imports first, so that cyclic static imports can be resolved correctly
+	index = resolveImports(numberOfStatements, resolvedImports, index, isNotStaticImport);
```

Comment:
This is ugly, but preserves the original code as much as possible. While still going over the non-static imports first, so that static imports can be properly resolved afterward.

The previous commit tried to do the same but less ugly, likely some check was not being done properly due to looping over incomplete imports somewhere in that code.

Reply:
Seems like tests pass now. @iloveeclipse can you review?